File: leetcode_ly/c++/List/AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* front = l1;
        ListNode* end = l2;
        ListNode* frontCopy;
        ListNode* endCopy;
        int addition = 0;
        while(front != NULL && end != NULL){
            frontCopy = front;
            endCopy = end;
            front->val = front->val + end->val + addition;
            addition = 0;
            if(front->val >= 10){
                addition = 1;
                front->val = front->val % 10;
            }
            front = front->next;
            end = end->next;
        }
        if(front == NULL && end != NULL){
            frontCopy->next = end;
            end->val = end->val + addition;
            while(end->val >= 10){
                frontCopy = end;
                addition = 1;
                end->val = end->val % 10;
                if(end->next){
                    end = end->next;
                    end->val = end->val + addition;
                }
                else{
                    ListNode* temp = new ListNode(1);
                    frontCopy->next = temp;
                }
            }
        }
        else if(end == NULL && front != NULL){
            frontCopy->next = front;
            front->val = front->val + addition;
            while(front && front->val >= 10){
                frontCopy = front;
                addition = 1;
                front->val = front->val % 10;
                if(front->next){
                    front = front->next;
                    front->val = front->val + addition;
                }
                else{
                    ListNode* temp = new ListNode(1);
                    frontCopy->next = temp;
                }
            }
        }
        else if(end == NULL && front == NULL && addition == 1){
            ListNode* temp = new ListNode(1);
            frontCopy->next = temp;
        }
        return l1;
    }
};

File: leetcode_ly/java/List/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1 == null) return l2;
        if(l2 == null) return l1;
        int len1 = getLen(l1);
        int len2 = getLen(l2);
        if(len1 < len2) return addTwoNumbers(l2,l1);
        
        int flag = 0;
        ListNode temp = l1;
        ListNode ptr = l1;
        while(l1 != null && l2 != null){
            l1.val += (l2.val + flag);
            flag = 0;
            if(l1.val >= 10){
                l1.val %= 10;
                flag = 1;
            }
            temp = l1;
            l1 = l1.next;
            l2 = l2.next;
        }
        while(l1 != null && flag == 1){
            temp = l1;
            int value = l1.val + flag;
            l1.val = (l1.val + flag) % 10;
            flag = value / 10;
            l1 = l1.next;
        }
        if(l1 == null && flag == 1){
            ListNode node = new ListNode(1);
            temp.next = node;
        }
        return ptr;
    }
    
    public int getLen(ListNode head){
        int len = 0;
        while(head != null){
            len++;
            head = head.next;
        }
        return len;
    }
}